ghost/link-tracking/lib/LinkClickTrackingService.js

```diff
@@ -78,7 +97,102 @@ class LinkClickTrackingService {
         });
     }
 
-    /** 
+    /**
+   * validate and manage the new redirect url in filter
+   * `to` url needs decoding and transformation to relative url for comparision
+   * @param {string} filter
+   * @returns {Object} parsed filter
+   * @throws {errors.BadRequestError}
+   */
+    #parseLinkFilter(filter) {
+        // decode filter to manage any encoded uri components
+        filter = decodeURIComponent(filter);
+
+        try {
+            const filterJson = nql(filter).parse();
+            const postId = filterJson?.$and?.[0]?.post_id;
+            const redirectUrl = new URL(filterJson?.$and?.[1]?.to);
+            if (!postId || !redirectUrl) {
+                throw new errors.BadRequestError({
+                    message: tpl(messages.invalidFilter)
+                });
+            }
+            return {
+                postId,
+                redirectUrl
+            };
+        } catch (e) {
+            throw new errors.BadRequestError({
+                message: tpl(messages.invalidFilter),
+                context: e.message
+            });
+        }
+    }
+
+    #getRedirectLinkWithAttribution({newLink, oldLink, postId}) {
+        const newUrl = new URL(newLink);
+        const oldUrl = new URL(oldLink);
+        // append newsletter ref query param from oldUrl to newUrl
+        if (oldUrl.searchParams.has('ref')) {
+            newUrl.searchParams.set('ref', oldUrl.searchParams.get('ref'));
+        }
+
+        // append post attribution to site urls
+        const isSite = this.#urlUtils.isSiteUrl(newUrl);
+        if (isSite) {
+            newUrl.searchParams.set('attribution_type', 'post');
+            newUrl.searchParams.set('attribution_id', postId);
+        }
+        return newUrl;
+    }
+
+    async #updateLinks(data, options) {
+        const filterOptions = _.pick(options, ['transacting', 'context', 'filter']);
+
+        // decode and parse filter to manage new redirect url
+        const {postId, redirectUrl} = this.#parseLinkFilter(filterOptions.filter);
+
+        // manages transformation of current url to relative for comparision
+        const transformedOldUrl = this.#urlUtils.absoluteToTransformReady(redirectUrl.href);
+        const filterQuery = `post_id:${postId}+to:'${transformedOldUrl}'`;
+
+        const updatedFilterOptions = {
+            ...filterOptions,
+            filter: filterQuery
+        };
+
+        // get new redirect link with proper attribution
+        const newRedirectUrl = this.#getRedirectLinkWithAttribution({
+            newLink: data.meta?.link?.to,
+            oldLink: redirectUrl.href,
+            postId
+        });
+
+        // fetch link rows with old redirect url to update
+        const linkRows = await this.#LinkRedirect.getFilteredCollectionQuery(updatedFilterOptions)
```

Comment:
I don't think we should be using a bookshelf model here directly, instead the link redirect service/repository (probably service) would be better ☺️

Reply:
Updated ✅